### ISSUE TITLE
Add an option to type English in Kaiboard compatible mode

### ISSUE THREAD

**Issue by alex-the-man** [Cantoboard/Cantoboard#14]

**Comment by sdugoten** [comment]
Just adding the Kaiboard behavior here.

When you type "佢英文名叫Lumd Pazk"

For kaiboard user, we will type "keui ying man meng giu Lumd Pazk [enter]"
For Cantoboard, I will have to type "keui ying man meng giu Lumd [enter] Pazk [enter]"

And I found two more issue

For example when you want type type the English word "luk", it will automatically change into "六" when you press space. I think it's regarding the same algorithm on inputting English word.

Another thing is that when you type "Cha" and then space in Kaiboard, it will show as "Cha" as we expected. If you change your mind and want to turn the word "Cha" into "差", you just need to backspace once, and the word list will show all the Chinese word pronounce as "cha". I think somehow Kaiboard will make a temp buffer to store the last typed word in case you need to change the English word back into Chinese.

**Comment by alex-the-man** [comment]
I will add a new space mode called "Pass-through input". It will insert the current input buffer as it's to the text field.
> After changing the behavior of English typing in Cantoboard, I think you can remove the listing of any English words in the selection list and just leaving the Chinese word in the selection.
The current design supports auto correct. Does auto correct work for you?

**Comment by sdugoten** [comment]
> > For kaiboard user, we will type "keui ying man meng giu Lumd Pazk [enter]"
> 
> Is tapping space required after entering Lumd in Kaiboard?

yes, tapping space is required, but pressing enter is not required.  Basically, the behavior is just like you regular type English sentence, you tap space in between words, and only select Chinese char on the selection list when needed.  So, you can type English sentence as you were using regular English keyboard.

> 
> > After changing the behavior of English typing in Cantoboard, I think you can remove the listing of any English words in the selection list and just leaving the Chinese word in the selection.
> 
> The current design supports auto correct. Does auto correct work for you?

The auto-correct works, but since you can type English sentence as you are using a regular English keyboard as I described above, you don't need to show any English words in the selection list.  In fact, I can use Kaiboard to type a whole paragraphs of English sentence just like how I use a regular English keyboard.

**Comment by alex-the-man** [comment]
> Basically, the whole idea is that the preferred behavior would be having the Cantoboard type English just like you are using a regular English keyboard. You may choose Chinese character on the selection list if you wish to, but the experience should work just like a English keybaord if you choose not to tap on any Chinese character on the selection list. Since the behavior works just like a regular English keyboard if you choose not to select any Chinese character on the selection list, then there is no need to prompt English word in the selection list, right?

Just to make sure I understand you correctly, I still show autocorrected English candidates in the candidate bar, but skip showing the exact match as the user could just tap space to enter it? 

I like your suggestion. Let me take a stab at it.

**Comment by sdugoten** [comment]
Hi, you can keep the autocorrected English candidates in the bar but skip showing the exact match.  However, I would suggest you to change the behavior of pressing space bar become auto selecting the candidate or moving the candidates bar to the next page.  Pressing space bar should always display a space just like what you would expect in a regular English keyboard.

The reason is that let say I need to type a word that your dictionary don't recognize , something like "abcde"  , the current behavior would forced me to press "enter" in order to have the "abcde" register into the text box.   

If you are using a regular English keyboard and type "abcde efghi", all you need to do is to naturally type "abcde <space> efghi"  
When you are using Canton board, due to the auto correction, when you type any word that is not known to a dictionary and you naturally pressing <space>, you end up having a auto corrected word which is not what you want.

Basically, the whole idea is mimicking a real English keyboard and it should type exactly like what you would expect in a English keyboard.  And the keyboard will allow you to tap on the Chinese char on the candidate bar and turn into a Chinese character.  

By doing this behavior change, you basically can use Cantoboard as a real English keyboard replacement , AND it will allow you to type Chinese as well.

**Comment by alex-the-man** [comment]
Understood. I'm working at this right now, I will update this issue when the test build is available on TestFlight.

**Comment by alex-the-man** [comment]
Could you please try the latest TestFlight build: 1.1.5 (3)?

**Comment by sdugoten** [comment]
This is almost perfect.  Wonderful job!

One last thing is having a buffer to store the last input.  For example...  I type "chat" and then a < space >.  I actually want to turn the word "chat" into "七".  So, I can backspace once, and the cursor is at the end of the word "chat" now, the selection box will display candidate for all word pronounce as "chat" such as 七, 漆, 柒...etc.

Other than that, I have no more issue on this.  I have a separate issue on combability input mapping which I will put into the other post

**Comment by alex-the-man** [comment]
> One last thing is having a buffer to store the last input. For example... I type "chat" and then a < space >. I actually want to turn the word "chat" into "七". So, I can backspace once, and the cursor is at the end of the word "chat" now, the selection box will display candidate for all word pronounce as "chat" such as 七, 漆, 柒...etc.

I think adding this correction behavior will confuse a lot of users. In this case, it's better to just retype the word. 

> And the "enter" key should just make it as "enter" instead of selection of candidate. For example, right now when you go to google.com and type "ip address", you will have to press "enter" twice in order for google to search.

This behavior doesn't match the default behavior of iOS keyboards. With default iOS Chinese input methods, users have to hit enter twice.

**Comment by sdugoten** [comment]
> > One last thing is having a buffer to store the last input. For example... I type "chat" and then a < space >. I actually want to turn the word "chat" into "七". So, I can backspace once, and the cursor is at the end of the word "chat" now, the selection box will display candidate for all word pronounce as "chat" such as 七, 漆, 柒...etc.
> 
> I think adding this correction behavior will confuse a lot of users. In this case, it's better to just retype the word.
> 
> > And the "enter" key should just make it as "enter" instead of selection of candidate. For example, right now when you go to google.com and type "ip address", you will have to press "enter" twice in order for google to search.
> 
> This behavior doesn't match the default behavior of iOS keyboards. With default iOS Chinese input methods, users have to hit enter twice.

The problem is that the behaviour is wierd, you try to open whatsapp, and then type something in the text box, something as simple as "OK" . As soon as you finish typing the character"K" and you press on the blue arrow send button on the right hand side, you can't send message no matter how many times you click on the send

**Comment by alex-the-man** [comment]
> The enter key problem is that the behaviour is wierd, you try to open whatsapp, and then type something in the text box, something as simple as "OK" . As soon as you finish typing the character"K" and you press on the blue arrow send button on the right hand side, you can't send message no matter how many times you click on the send button

I share the same frustration. The implementation of iOS's marked text API for third party keyboard extensions is horribly implemented. The only way around is to ditch iOS's marked text API but that requires a lot of changes which I'm not ready for at the moment.

> Another strange behaviour tied to this Enter key behavior is that when you try to use Cantoboard to login to github.com. In the username textbox, I type in my username"sdugoten" and then tap on the password field, the word "sdugoten" will magically disappear because it didn't register without pressing the enter key. The behavior is just not natural .

This looks like buggy interaction between the marked text API and the keyboard extension. I will investigate.

**Comment by alex-the-man** [comment]
> Another strange behaviour tied to this Enter key behavior is that when you try to use Cantoboard to login to github.com. In the username textbox, I type in my username"sdugoten" and then tap on the password field, the word "sdugoten" will magically disappear because it didn't register without pressing the enter key. The behavior is just not natural .

Which browser did you use?

**Comment by sdugoten** [comment]
https://streamable.com/9dd0ig

Safari

**Comment by alex-the-man** [comment]
Ditching marked text is a very big change. But apps using Flutter has serious compatibility issue with marked text. I might add immediate mode support in the future.

**Comment by alex-the-man** [comment]
#26
